pkg/storage/etcd3/etcdwatcher/watcher.go

```diff
@@ -0,0 +1,366 @@
+	return &watcher{
+		client:    client,
+		codec:     codec,
+		versioner: versioner,
+	}
+}
+
+func (w *watcher) Watch(ctx context.Context, key string, rev int64, recursive bool, filter storage.FilterFunc) (watch.Interface, error) {
+	if recursive && !strings.HasSuffix(key, "/") {
+		key += "/"
+	}
+	wc := &watchChan{
+		watcher:           w,
+		key:               key,
+		initialRev:        rev,
+		recursive:         recursive,
+		filter:            filter,
+		resultChan:        make(chan watch.Event, 1),
+		incomingEventChan: make(chan *event, 1),
+		errChan:           make(chan error, 1),
+	}
+	wc.ctx, wc.cancel = context.WithCancel(ctx)
+	go wc.run()
+	return wc, nil
+}
+
+func (wc *watchChan) run() {
+	go wc.startWatching()
+
+	var doneWithResultChan sync.WaitGroup
+	doneWithResultChan.Add(1)
+	go wc.eventProcessor(&doneWithResultChan)
+
+	var errResult *watch.Event
+	select {
+	case err := <-wc.errChan:
+		errResult = wc.parseError(err)
+		wc.cancel()
+	case <-wc.ctx.Done():
+	}
+	// error result is guaranteed to be received by user before closing ResultChan.
+	if errResult != nil {
+		wc.resultChan <- *errResult
+	}
+	// we need to wait until resultChan wouldn't be sent to anymore
+	doneWithResultChan.Wait()
+	close(wc.resultChan)
+}
+
+func (wc *watchChan) Stop() {
+	wc.cancel()
+}
+
+func (wc *watchChan) ResultChan() <-chan watch.Event {
+	return wc.resultChan
+}
+
+// sync tries to retrieve existing data and send them to process.
+// The revision to watch will be set to the revision in response.
+func (wc *watchChan) sync() error {
+	opts := []clientv3.OpOption{}
+	if wc.recursive {
+		opts = append(opts, clientv3.WithPrefix())
+	}
+	getResp, err := wc.watcher.client.Get(wc.ctx, wc.key, opts...)
+	if err != nil {
+		return err
+	}
+	wc.initialRev = getResp.Header.Revision
+
+	for _, kv := range getResp.Kvs {
+		wc.sendEvent(parseEtcdKV(kv))
+	}
+	return nil
+}
+
+// startWatching does:
+// - wait for sync to complete
+// - watch on given key and send events to process.
+func (wc *watchChan) startWatching() {
+	if wc.initialRev != 0 {
+		err := wc.sync()
+		if err != nil {
+			wc.sendError(err)
+			return
+		}
+	}
+	opts := []clientv3.OpOption{clientv3.WithRev(wc.initialRev + 1)}
+	if wc.recursive {
+		opts = append(opts, clientv3.WithPrefix())
+	}
+	wch := wc.watcher.client.Watch(wc.ctx, wc.key, opts...)
+	for wres := range wch {
+		if wres.Err() != nil {
+			// If there is an error on server (e.g. compaction), the channel will return it before closed.
+			wc.sendError(wres.Err())
+			return
+		}
+		for _, e := range wres.Events {
+			wc.sendEvent(parseEtcdEvent(e))
+		}
+	}
+}
+
+// eventProcessor receives events from etcd watcher and
+// stores them in local queue if deserializer couldn't deserialize objects fast enough.
+// In this way, we won't block etcd watcher from receiving events.
+func (wc *watchChan) eventProcessor(wg *sync.WaitGroup) {
+	eventChan := make(chan *event, 100)
+	go wc.deserializer(wg, eventChan)
```

Comment:
the usage of wg is confusing here. i think eventProcessor should wait its children, not run. run should wait on eventProcessor.